packages/promises/promise.pony

```diff
@@ -69,6 +69,161 @@ actor Promise[A: Any #share]
+
+    Take a look at what happens in the example, when we get to
+    `PrintIssueTitle`, we can't print anything because we "don't have anything".
+    In order to print the issue title, we need an `Issue` not a
+    `Promise[Issue]`.
+
+    We could solve this by doing something like this:
+
+    ```pony
+    primitive PrintIssueTitle
+      fun apply(out: OutStream, issue: Promise[Issue]) =>
+        issue.next[None](ActuallyPrintIssueTitle~apply(out))
+
+    primitive ActuallyPrintIssueTitle
+      fun apply(out: OutStream, issue: Issue) =>
+        out.print(issue.title())
+    ```
+
+    That will work, however, it is kind of awful. When looking at:
+
+    ```pony
+        let repo: Promise[Repo] =
+          GitHub("my token").get_repo("ponylang/ponyc")
+        let issue = Promise[Promise[Issue]] =
+          repo.next[Promise[Issue]](FetchIssue~apply(1))
+        issue.next[None](PrintIssueTitle~apply(env.out))
+    ```
+
+    it can be hard to follow what is going on. We can only tell what is
+    happening because we gave `PrintIssueTitle` a very misleading name; it
+    doesn't print an issue title.
+
+    `flatten_next` addresses the problem of "we want the `Issue`, not the
+    intermediate `Promise`". `flatten_next` takes an intermediate promise and
+    unwraps it into the fulfilled type. You get to write your promise chain
+    without having to worry about intermediate promises.
+
+    Updated to use `flatten_next`, our API example becomes:
+
+    ```pony
+    actor Main
+      new create(env: Env) =>
+        let repo: Promise[Repo] =
+          GitHub("my token").get_repo("ponylang/ponyc")
+
+        let issue = Promise[Issue] =
+          repo.flatten_next[Issue](FetchIssue~apply(1))
+
+        issue.next[None](PrintIssueTitle~apply(env.out))
+
+    primitive FetchIssue
+      fun apply(number: I64, repo: Repo): Promise[Issue] =>
+        repo.get_issue(number)
+
+    primitive PrintIssueTitle
+      fun apply(out: OutStream, issue: Issue) =>
+        out.print(issue.title())
+    ```
+
+    Our promise `Issue`, is no longer a `Promise[Promise[Issue]]`. By using
+    `flatten_next`, we have a much more manageable `Promise[Issue]` instead.
+
+    Other than unwrapping promises for you, `flatten_next` otherwise acts the
+    same as `next` so all the same rules apply to fulfillment and rejection.
+    """
+    let outer = Promise[B]
+
+    next[None](object iso
+      var f: Fulfill[A, Promise[B]] = consume fulfill
+      var r: Reject[Promise[B]] = consume rejected
+      let p: Promise[B] = outer
+
+      fun ref apply(value: A) =>
+        let fulfill' = f = {(value: A) => Promise[B]}
+        let rejected' = r = RejectAlways[Promise[B]]
+
+        try
+          let inner = try
+            (consume fulfill').apply(value)?
+          else
+            (consume rejected').apply()?
+          end
+
+          inner.next[None](
+              {(fulfilled: B) =>
+                p(fulfilled)
+              })
+        else
```

Comment:
@SeanTAllen - I think something like this is the right fix that @jasoncarr0 was pointing out.

```suggestion
              {(fulfilled: B) => p(fulfilled) },
              {() : B? => p.reject(); error })
        else
```

Reply:
that sadly doesnt compile

Reply:
/home/sean/code/ponylang/ponyc/packages/promises/promise.pony:226:13: argument not a subtype of parameter
            {() : B? => p.reject(); error})
            ^
    Info:
    /home/sean/code/ponylang/ponyc/packages/promises/promise.pony:226:13: argument type is {(): B ?}[A #share, B #share] iso^
                {() : B? => p.reject(); error})
                ^
    /home/sean/code/ponylang/ponyc/packages/promises/promise.pony:48:5: parameter type is Reject[None val] iso^
        rejected: Reject[B] = RejectAlways[B])
        ^

Reply:
fixed

Reply:
This compiles:

```pony
            {(fulfilled: B) => p(fulfilled)},
            {()? => p.reject(); error})
```